### DISCUSSION TITLE
App crash when deleting images from canvas

### DISCUSSION THREAD

**Discussion by Icyretsz** [fabricjs/fabric.js#10088] [Q&A]
The error is as follows: TypeError: Cannot read properties of undefined (reading 'getRetinaScaling')
It doesn't happen on other objects like shapes for textboxes, just happens on images objects.

https://github.com/user-attachments/assets/03ddc785-fafe-404a-99f5-04815ad7ea1d

Below is my implementation of the delete method:
`deleteObject: () => {

                canvas.getActiveObjects().forEach((object) => {

                    object.off()

                    canvas.remove(object);

                    canvas.discardActiveObject();

                })

                canvas.fire('object:modified')

                canvas.renderAll();

            }`
And here is my implementation of adding an image:
`addMedia: (url : string) => {

                fabric.Image.fromURL(url, function(oImg) {

                    oImg.scale(0.1)

                    canvas.add(oImg);

                    addProc(oImg); //this is just a method for centering and set active object

                });

            }`

**Comment by Icyretsz** [comment]
It seems like I have to manually define canvas operations after adding object directly inside the callback funtion when adding image. Using pre-defined function won't work.

For reference, my addProc method:
`const addProc = (object: fabric.Object) => {
            center(object);
            canvas.add(object);
            const currentId : string = String(canvas.getObjects().length - 1)
            object.set('name', currentId)
            object.on('mousedown', function(event) {
                if(event.button === 3) {
                    console.log('right click')
                }
            })
            canvas.fire('object:modified')
            canvas.setActiveObject(object);
            canvas.renderAll();
        };`
        Using addProc inside the callback won't work

        I'll have to define centerObjects, setActiveObjects.... directly inside the callback function
        `addMedia: (url : string) => {
                fabric.Image.fromURL(url, function(oImg) {
                    const currentId : string = String(canvas.getObjects().length - 1)
                    oImg.set('name', currentId)
                    oImg.scale(0.1)
                    canvas.add(oImg);
                    canvas.setActiveObject(oImg);
                    canvas.centerObject(oImg);
                    canvas.renderAll();
                    oImg.on('mousedown', function(event) {
                        if(event.button === 3) {
                            console.log('right click')
                        }
                    })
                });
                canvas.fire('object:modified')
            }`

**Reply by asturur** [reply]
To center an object or to set it as the active one, yes that has to be in the canvas already.
to scale, set a name or add a mousedown event it doesn't matter.

I higly suggest you switch to fabric 6 since fabric 5 is unmaintained

**Reply by asturur** [reply]
Also your implementation of delete object seems a bit complicated, removing an object from canvas should be all you need, then if you don't need the object anymore you can call object.dispose()
